### ISSUE TITLE
Timeout for process termination upon receiving sigterm should be a user configurable parameter

### ISSUE THREAD

**Issue by Shikib** [pytorch/pytorch#119856]
### 🚀 The feature, motivation and pitch

I'm trying to implement graceful exit on SLURM pre-emption. SLURM gives 10 minutes upon pre-emption (i.e., SIGTERM; sleep 600; SIGKILL), but it seems that torchrun will send a SIGKILL from the launcher to all the child processes after 30 seconds (based on [this PR](https://github.com/pytorch/pytorch/pull/61602)). Since graceful termination (e.g., saving checkpoint) can take >30seconds when working with large models, the default 30s timeout should be a user-configurable parameter.

Minimal example/reproduction of this issue:

```
import time
from pathlib import Path
import signal

def ignore(s,f):
    print('!')

signal.signal(signal.SIGTERM, ignore)

p = Path("/tmp/finished")
p.touch()
p.unlink()
time.sleep(200)
p.touch()
```

1. Launch with `python -u -m accelerate.commands.launch --num_processes 1 test.py`
2. Send sigterm `kill -TERM <pid>`
3. Process receives `SIGKILL from launcher (torch.distributed.elastic.multiprocessing.api: [WARNING] Unable to shutdown process 719448 via Signals.SIGTERM, forcefully exiting via Signals.SIGKILL)` -- from this [line](https://github.com/pytorch/pytorch/blame/cd51d2a3ecc8ac579bee910f6bafe41a4c41ca80/torch/distributed/elastic/multiprocessing/api.py#L551)

Tagging based on Slack discussion: @kurman @wconstab

**Comment by stas00** [comment]
yes, accelerate will have to expose this new flag as it manually forwards the flags to torchrun.

but I think at this point we are just asking for torchrun to implement that and then will ask accelerate to support it as well.

**Comment by collinmccarthy** [comment]
I'm very interested in this feature as well. In my scenario the cleanup involves syncing checkpoints to W&B which takes at least a minute or two, so I'd like to allow up to 5 minutes between SIGTERM and SIGKILL. 

Is there any work around right now, maybe by setting `monitor_interval`? Doesn't seem like that would give me what I want though, and have other potential downsides.

**Comment by TJ-Solergibert** [comment]
Facing the same issue. With `torchrun` I can't gracefully exit my application as I need some minutes to store a checkpoint. Any hints?

Thanks

**Comment by kabouzeid** [comment]
Same here. SIGUSR is not handled and SIGTERM kills very quickly with non-configurable timeout. related: https://github.com/pytorch/pytorch/issues/60511

**Comment by dwan-ith** [comment]
Hi,
I noticed that #172596 implemented the configurable shutdown timeout for signal-triggered shutdown and final cleanup.
On current main, LocalElasticAgent._stop_workers() still calls _shutdown() without passing self._shutdown_timeout. hence, worker-stop paths used during failures and elastic restarts fall back to the 30-second default, even when the user configured another value.

I have a small follow-up patch that changes this to:
self._shutdown(timeout=self._shutdown_timeout)

and adds a regression test covering _stop_workers().

One tradeoff is that a long user-configured timeout would then also apply during restart-triggered shutdowns, (potentially) delaying elastic recovery from unhealthy workers. The current CLI documentation describes this as the graceful worker-shutdown timeout generally, but I could not find documentation indicating whether restart paths should intentionally retain 30 seconds.
Is applying the configured timeout to _stop_workers() the remaining work intended by the recently added good first issue label? 

If so, I’d be happy to open the follow-up PR.
